File: lib/product/file_adapters.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'crypto_utils.dart';

enum FileAdapterTier { native, sandboxedCore, plugin }

enum FileAdapterCapability {
  detect,
  inspect,
  extract,
  preview,
  create,
  transform,
  validate,
}

class FileAdapterDescriptor {
  const FileAdapterDescriptor({
    required this.id,
    required this.tier,
    required this.extensions,
    required this.mediaTypes,
    required this.capabilities,
    this.sandboxRequired = false,
  });

  final String id;
  final FileAdapterTier tier;
  final Set<String> extensions;
  final Set<String> mediaTypes;
  final Set<FileAdapterCapability> capabilities;
  final bool sandboxRequired;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'id': id,
        'tier': tier.name,
        'sandboxRequired': sandboxRequired,
        'extensions': extensions.toList()..sort(),
        'mediaTypes': mediaTypes.toList()..sort(),
        'capabilities': capabilities.map((item) => item.name).toList()..sort(),
      };
}

class FileInspectionResult {
  const FileInspectionResult({
    required this.adapter,
    required this.path,
    required this.sizeBytes,
    required this.sha256,
    required this.metadata,
  });

  final FileAdapterDescriptor adapter;
  final String path;
  final int sizeBytes;
  final String sha256;
  final Map<String, dynamic> metadata;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'adapter': adapter.toJson(),
        'path': path,
        'sizeBytes': sizeBytes,
        'sha256': sha256,
        'metadata': metadata,
      };
}

class FileAdapterRegistry {
  const FileAdapterRegistry();

  List<FileAdapterDescriptor> get all => const <FileAdapterDescriptor>[
        FileAdapterDescriptor(
          id: 'text',
          tier: FileAdapterTier.native,
          extensions: <String>{'txt', 'md', 'log'},
          mediaTypes: <String>{'text/plain', 'text/markdown'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'json',
          tier: FileAdapterTier.native,
          extensions: <String>{'json'},
          mediaTypes: <String>{'application/json'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'yaml',
          tier: FileAdapterTier.native,
          extensions: <String>{'yaml', 'yml'},
          mediaTypes: <String>{'application/yaml', 'text/yaml'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'xml',
          tier: FileAdapterTier.native,
          extensions: <String>{'xml'},
          mediaTypes: <String>{'application/xml', 'text/xml'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'csv',
          tier: FileAdapterTier.native,
          extensions: <String>{'csv'},
          mediaTypes: <String>{'text/csv'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'image',
          tier: FileAdapterTier.native,
          extensions: <String>{'png', 'jpg', 'jpeg', 'gif', 'webp'},
          mediaTypes: <String>{'image/png', 'image/jpeg', 'image/gif', 'image/webp'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'zip',
          tier: FileAdapterTier.native,
          extensions: <String>{'zip'},
          mediaTypes: <String>{'application/zip'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
        ),
        FileAdapterDescriptor(
          id: 'pdf',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'pdf'},
          mediaTypes: <String>{'application/pdf'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
        FileAdapterDescriptor(
          id: 'ooxml',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'docx', 'xlsx', 'pptx'},
          mediaTypes: <String>{
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            'application/vnd.openxmlformats-officedocument.presentationml.presentation',
          },
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
        FileAdapterDescriptor(
          id: 'opendocument',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'odt'},
          mediaTypes: <String>{'application/vnd.oasis.opendocument.text'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
        FileAdapterDescriptor(
          id: 'rtf',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'rtf'},
          mediaTypes: <String>{'application/rtf', 'text/rtf'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
        FileAdapterDescriptor(
          id: 'epub',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'epub'},
          mediaTypes: <String>{'application/epub+zip'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
        FileAdapterDescriptor(
          id: 'email',
          tier: FileAdapterTier.sandboxedCore,
          extensions: <String>{'eml'},
          mediaTypes: <String>{'message/rfc822'},
          capabilities: <FileAdapterCapability>{
            FileAdapterCapability.detect,
            FileAdapterCapability.inspect,
            FileAdapterCapability.extract,
            FileAdapterCapability.preview,
            FileAdapterCapability.validate,
          },
          sandboxRequired: true,
        ),
      ];

  FileAdapterDescriptor detect(File file) {
    final extension = file.path.contains('.')
        ? file.path.split('.').last.toLowerCase()
        : '';
    for (final adapter in all) {
      if (adapter.extensions.contains(extension)) {
        return adapter;
      }
    }
    return const FileAdapterDescriptor(
      id: 'binary',
      tier: FileAdapterTier.plugin,
      extensions: <String>{},
      mediaTypes: <String>{'application/octet-stream'},
      capabilities: <FileAdapterCapability>{
        FileAdapterCapability.detect,
        FileAdapterCapability.inspect,
      },
      sandboxRequired: true,
    );
  }

  Future<FileInspectionResult> inspect(File file) async {
    final bytes = await file.readAsBytes();
    final adapter = detect(file);
    final metadata = <String, dynamic>{
      'extension': file.path.contains('.')
          ? file.path.split('.').last.toLowerCase()
          : '',
    };
    if (adapter.id == 'image' &&
        bytes.length >= 24 &&
        bytes[0] == 0x89 &&
        bytes[1] == 0x50 &&
        bytes[2] == 0x4E &&
        bytes[3] == 0x47) {
      final view = ByteData.sublistView(Uint8List.fromList(bytes), 16, 24);
      metadata['width'] = view.getUint32(0, Endian.big);
      metadata['height'] = view.getUint32(4, Endian.big);
    }
    if (adapter.id == 'email') {
      final text = utf8.decode(bytes, allowMalformed: true);
      final subject = RegExp(r'^subject:\s*(.+)$', caseSensitive: false, multiLine: true)
              .firstMatch(text)
              ?.group(1) ??
          '';
      metadata['subject'] = subject.trim();
    }
    return FileInspectionResult(
      adapter: adapter,
      path: file.path,
      sizeBytes: bytes.length,
      sha256: Sha256.hex(bytes),
      metadata: metadata,
    );
  }

  Future<Map<String, dynamic>> validate(File file) async {
    final bytes = await file.readAsBytes();
    final adapter = detect(file);
    try {
      switch (adapter.id) {
        case 'text':
          utf8.decode(bytes);
          break;
        case 'json':
          jsonDecode(utf8.decode(bytes));
          break;
        case 'yaml':
          final text = utf8.decode(bytes);
          if (text.trim().isEmpty) {
            throw StateError('YAML is empty.');
          }
          break;
        case 'xml':
          final text = utf8.decode(bytes);
          if (!text.trimLeft().startsWith('<')) {
            throw StateError('XML does not start with an element.');
          }
          break;
        case 'csv':
          final text = utf8.decode(bytes);
          if (!text.contains(',') && !text.contains('\n')) {
            throw StateError('CSV is not delimited text.');
          }
          break;
        case 'image':
          final isPng = bytes.length >= 4 &&
              bytes[0] == 0x89 &&
              bytes[1] == 0x50 &&
              bytes[2] == 0x4E &&
              bytes[3] == 0x47;
          final isJpeg =
              bytes.length >= 3 && bytes[0] == 0xFF && bytes[1] == 0xD8 && bytes[2] == 0xFF;
          final signature = utf8.decode(bytes.take(6).toList(), allowMalformed: true);
          if (!(isPng || isJpeg || signature.startsWith('GIF8') || signature.startsWith('RIFF'))) {
            throw StateError('Image signature is not recognized.');
          }
          break;
        case 'zip':
        case 'ooxml':
        case 'opendocument':
        case 'epub':
          final header = bytes.length >= 4
              ? bytes.take(4).toList()
              : const <int>[];
          final isZip = header.length == 4 &&
              header[0] == 0x50 &&
              header[1] == 0x4B &&
              header[2] == 0x03 &&
              header[3] == 0x04;
          if (!isZip) {
            throw StateError('ZIP container header is missing.');
          }
          break;
        case 'pdf':
          final text = utf8.decode(bytes, allowMalformed: true);
          if (!text.startsWith('%PDF-') || !text.contains('%%EOF')) {
            throw StateError('PDF markers are missing.');
          }
          break;
        case 'rtf':
          final text = utf8.decode(bytes, allowMalformed: true);
          if (!text.trimLeft().startsWith(r'{\rtf')) {
            throw StateError('RTF header is missing.');
          }
          break;
        case 'email':
          final text = utf8.decode(bytes, allowMalformed: true).toLowerCase();
          if (!(text.contains('\nsubject:') ||
              text.contains('\nfrom:') ||
              text.contains('\ndate:') ||
              text.startsWith('from:'))) {
            throw StateError('Email headers are missing.');
          }
          break;
      }
      return <String, dynamic>{'passed': true, 'adapterId': adapter.id};
    } catch (error) {
      return <String, dynamic>{
        'passed': false,
        'adapterId': adapter.id,
        'error': '$error',
      };
    }
  }
}

File: tool/file_adapters.py
#!/usr/bin/env python3
"""Deterministic core file-adapter registry for Kristin v1.8."""
from __future__ import annotations

from dataclasses import dataclass
import email
import hashlib
import json
from pathlib import Path
import struct
import zipfile
import xml.etree.ElementTree as ET

VERSION = "1.9.0+190"
PNG_SIGNATURE = bytes.fromhex('89504e470d0a1a0a')
JPEG_SIGNATURE = bytes.fromhex('ffd8ff')
GIF_SIGNATURES = (b'GIF87a', b'GIF89a')
WEBP_SIGNATURE = b'RIFF'


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


@dataclass(frozen=True)
class FileAdapter:
    id: str
    tier: str
    extensions: tuple[str, ...]
    media_types: tuple[str, ...]
    capabilities: tuple[str, ...]
    sandbox_required: bool = False


BUILTINS: tuple[FileAdapter, ...] = (
    FileAdapter('text', 'native', ('txt', 'md', 'log'), ('text/plain', 'text/markdown'), ('detect', 'inspect', 'extract', 'preview', 'validate')),
    FileAdapter('json', 'native', ('json',), ('application/json',), ('detect', 'inspect', 'extract', 'preview', 'validate')),
    FileAdapter('yaml', 'native', ('yaml', 'yml'), ('application/yaml', 'text/yaml'), ('detect', 'inspect', 'extract', 'preview', 'validate')),
    FileAdapter('xml', 'native', ('xml',), ('application/xml', 'text/xml'), ('detect', 'inspect', 'extract', 'preview', 'validate')),
    FileAdapter('csv', 'native', ('csv',), ('text/csv',), ('detect', 'inspect', 'extract', 'preview', 'validate')),
    FileAdapter('image', 'native', ('png', 'jpg', 'jpeg', 'gif', 'webp'), ('image/png', 'image/jpeg', 'image/gif', 'image/webp'), ('detect', 'inspect', 'preview', 'validate')),
    FileAdapter('zip', 'native', ('zip',), ('application/zip',), ('detect', 'inspect', 'preview', 'validate')),
    FileAdapter('pdf', 'sandboxed_core', ('pdf',), ('application/pdf',), ('detect', 'inspect', 'preview', 'validate'), sandbox_required=True),
    FileAdapter('ooxml', 'sandboxed_core', ('docx', 'xlsx', 'pptx'), ('application/vnd.openxmlformats-officedocument.wordprocessingml.document', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', 'application/vnd.openxmlformats-officedocument.presentationml.presentation'), ('detect', 'inspect', 'preview', 'validate'), sandbox_required=True),
    FileAdapter('opendocument', 'sandboxed_core', ('odt',), ('application/vnd.oasis.opendocument.text',), ('detect', 'inspect', 'preview', 'validate'), sandbox_required=True),
    FileAdapter('rtf', 'sandboxed_core', ('rtf',), ('application/rtf', 'text/rtf'), ('detect', 'inspect', 'extract', 'preview', 'validate'), sandbox_required=True),
    FileAdapter('epub', 'sandboxed_core', ('epub',), ('application/epub+zip',), ('detect', 'inspect', 'preview', 'validate'), sandbox_required=True),
    FileAdapter('email', 'sandboxed_core', ('eml',), ('message/rfc822',), ('detect', 'inspect', 'extract', 'preview', 'validate'), sandbox_required=True),
)


def registry_document() -> dict[str, object]:
    return {
        'schemaVersion': '1.0.0',
        'adapters': [
            {
                'id': adapter.id,
                'tier': adapter.tier,
                'sandboxRequired': adapter.sandbox_required,
                'extensions': list(adapter.extensions),
                'mediaTypes': list(adapter.media_types),
                'capabilities': list(adapter.capabilities),
            }
            for adapter in BUILTINS
        ],
    }


def detect(path: Path) -> FileAdapter:
    ext = path.suffix.lower().lstrip('.')
    for adapter in BUILTINS:
        if ext in adapter.extensions:
            return adapter
    return FileAdapter('binary', 'plugin', (ext,), (), ('detect', 'inspect'), sandbox_required=True)


def _png_dimensions(data: bytes) -> tuple[int, int] | None:
    if data.startswith(PNG_SIGNATURE) and len(data) >= 24:
        return struct.unpack('>II', data[16:24])
    return None


def inspect(path: Path) -> dict[str, object]:
    adapter = detect(path)
    data = path.read_bytes()
    info: dict[str, object] = {
        'path': str(path),
        'adapterId': adapter.id,
        'tier': adapter.tier,
        'sandboxRequired': adapter.sandbox_required,
        'sizeBytes': len(data),
        'sha256': sha256_hex(data),
        'extension': path.suffix.lower(),
    }
    if adapter.id == 'image':
        dims = _png_dimensions(data)
        if dims:
            info['width'], info['height'] = dims
    if adapter.id in {'zip', 'ooxml', 'opendocument', 'epub'}:
        with zipfile.ZipFile(path, 'r') as archive:
            info['memberCount'] = len(archive.namelist())
            info['members'] = archive.namelist()[:20]
    if adapter.id == 'email':
        message = email.message_from_bytes(data)
        info['subject'] = message.get('Subject', '')
    return info


def validate(path: Path) -> tuple[bool, str]:
    adapter = detect(path)
    data = path.read_bytes()
    try:
        if adapter.id == 'text':
            data.decode('utf-8')
            return True, 'UTF-8 text reopened successfully'
        if adapter.id == 'json':
            json.loads(data.decode('utf-8'))
            return True, 'JSON parsed successfully'
        if adapter.id == 'yaml':
            text = data.decode('utf-8')
            if not text.strip():
                raise ValueError('empty yaml')
            return True, 'YAML text is non-empty and UTF-8 decodable'
        if adapter.id == 'xml':
            ET.fromstring(data.decode('utf-8'))
            return True, 'XML parsed successfully'
        if adapter.id == 'csv':
            text = data.decode('utf-8')
            if ',' not in text and '\n' not in text:
                raise ValueError('not csv-like')
            return True, 'CSV reopened as UTF-8 text'
        if adapter.id == 'image':
            if data.startswith(PNG_SIGNATURE) or data.startswith(JPEG_SIGNATURE) or data[:6] in GIF_SIGNATURES or data.startswith(WEBP_SIGNATURE):
                return True, 'image signature recognized'
            raise ValueError('unknown image signature')
        if adapter.id == 'zip':
            with zipfile.ZipFile(path, 'r') as archive:
                if archive.testzip() is not None:
                    raise ValueError('zip corruption detected')
            return True, 'ZIP reopened successfully'
        if adapter.id == 'pdf':
            if not data.startswith(b'%PDF-'):
                raise ValueError('missing PDF header')
            if b'%%EOF' not in data[-2048:]:
                raise ValueError('missing PDF EOF marker')
            return True, 'PDF header and EOF marker detected'
        if adapter.id == 'ooxml':
            with zipfile.ZipFile(path, 'r') as archive:
                names = set(archive.namelist())
                if '[Content_Types].xml' not in names:
                    raise ValueError('OOXML content types missing')
                suffix = path.suffix.lower()
                if suffix == '.docx' and not any(name.startswith('word/') for name in names):
                    raise ValueError('DOCX word tree missing')
                if suffix == '.xlsx' and not any(name.startswith('xl/') for name in names):
                    raise ValueError('XLSX xl tree missing')
                if suffix == '.pptx' and not any(name.startswith('ppt/') for name in names):
                    raise ValueError('PPTX ppt tree missing')
            return True, 'OOXML package reopened successfully'
        if adapter.id == 'opendocument':
            with zipfile.ZipFile(path, 'r') as archive:
                mimetype = archive.read('mimetype').decode('utf-8')
                if not mimetype.startswith('application/vnd.oasis.opendocument'):
                    raise ValueError('OpenDocument mimetype missing')
            return True, 'OpenDocument package reopened successfully'
        if adapter.id == 'rtf':
            if not data.lstrip().startswith(b'{\\rtf'):
                raise ValueError('RTF header missing')
            return True, 'RTF header recognized'
        if adapter.id == 'epub':
            with zipfile.ZipFile(path, 'r') as archive:
                mimetype = archive.read('mimetype').decode('utf-8')
                if mimetype != 'application/epub+zip':
                    raise ValueError('EPUB mimetype mismatch')
            return True, 'EPUB package reopened successfully'
        if adapter.id == 'email':
            message = email.message_from_bytes(data)
            if not (message.get('From') or message.get('Subject') or message.get('Date')):
                raise ValueError('email headers missing')
            return True, 'RFC822 message parsed successfully'
    except Exception as exc:  # noqa: BLE001
        return False, f'{type(exc).__name__}: {exc}'
    return False, 'unsupported adapter'
